### ISSUE TITLE
Packaging the page elements in a 'page' class using a template?

### ISSUE THREAD

**Issue by csgregg** [tfry-git/EmbAJAX#16]
Very` useful library - thanks!

I wanting to create a 'page' class for each page which contains all the elements. I've been experimenting with creating a 'page' template class. Not making much progress. Any ideas or interesting in including something like this in the library?

Thanks

Relevant code snippet -

```
// Simplify to just id
class newbtn : public EmbAJAXCheckButton {
    public:
        newbtn(const char* id) : EmbAJAXCheckButton(id, id) {};
};

// For the helpers
template<size_t idx, typename T>
struct GetHelper;


// Build the page class template
template<typename ... Ttitle>
struct DataStructure
{
};

template<typename Ttitle, typename ... Telements>
struct DataStructure<Ttitle, Telements ...>
{
    DataStructure(const Ttitle& ptitle, const Telements& ... pelements)
        : title(ptitle)
        , elements(pelements...)
        , page_elements{&pelements...}
    //  , page(page_elements, ptitle)                    // This is causing trouble
    {}
    
    Ttitle title;
    DataStructure<Telements ... > elements;

    const EmbAJAXBase* page_elements[sizeof...(Telements)];

    //   EmbAJAXPage<(sizeof...(Telements)/sizeof(EmbAJAXBase*))> page;                    // This is causing trouble

    template<size_t idx>
    auto get() {  return GetHelper<idx, DataStructure<Ttitle,Telements...>>::get(*this); }
};


// Helpers to get to the elements
template<typename Ttitle, typename ... Telements>
struct GetHelper<0, DataStructure<Ttitle, Telements ... >>
{
    static Ttitle get(DataStructure<Ttitle, Telements...>& data) { return data.title; }
};

template<size_t idx, typename Ttitle, typename ... Telements>
struct GetHelper<idx, DataStructure<Ttitle, Telements ... >>
{
    static auto get(DataStructure<Ttitle, Telements...>& data) { return GetHelper<idx-1, DataStructure<Telements ...>>::get(data.elements); }
};


// Create the page
DataStructure<char*, newbtn, EmbAJAXMutableSpan> data("Title","check","check_d");

// Page handler
void datahandle() {
    data.get<2>().setValue(data.get<1>().isChecked() ? " Checked" : " Not checked");
};
```

**Comment by tfry-git** [comment]
Hi!

You mean you want to keep several pages in one array / list, but the compiler won't let you, because they have different sizes?

The easiest way out should be to simply create all your pages as ```EmbAJAXPage<1>```-instances, with their only child each being an ```EmbAJAXContainer<n>```-instance containing the real child-elements. Does this solve your problem, or am I missing part of what you're trying to do?

Or is it simply that you are trying to put an ```EmbAJAXPage``` as a member into another template class? In this case, I think all you will have to do is to repeat the ```NUM``` template parameter in the outer template. Like:

```
template<typename Ttitle, typename ... Telements, size_t NUM> struct DataStructure<Ttitle, Telements ...>
{
    DataStructure(const Ttitle& ptitle, const Telements& ... pelements)
        : title(ptitle)
        , elements(pelements...)
        , page_elements{&pelements...}
        , page(page_elements, ptitle)                    // This is causing trouble
    {}
    
    Ttitle title;
    DataStructure<Telements ... > elements;

    const EmbAJAXBase* page_elements<NUM>;

   EmbAJAXPage<NUM> page;  
```

I admit this is not terribly elegant, but could be a way out.

A hybrid solution could be to implement a cousin to ```EmbAJAXContainer```. It could be almost identical, but with the template parameter converted to a regular member variable, instead(*). Then you can use that in an ```EmbAJAXPage<1>``` to hold the actual page elements.

--

(*) In fact, I *am* starting to think that I have been overdoing it with the templating at this point, and this is how I should have designed EmbAJAXContainer in the first place. It's a bit difficult to change, now, however, without breaking things.

**Comment by tfry-git** [comment]
This: https://github.com/tfry-git/EmbAJAX/commit/6c0aec41a40d973e7eedadfe3a6e69fd48c4734d
is roughly what I have in mind for that cousin to EmbAJAXContainer. I'm not quite satisfied with it, yet, but it would probably let you solve your problem.

**Comment by csgregg** [comment]
Hi Thomas - thanks for the quick response. Back story on what I am wanting to do - I want to put all the elements on each page in one class per page, and then create an array of pointers to the pages, to build a 'website'. I'm going to serve the pages from html files in SPIFFS built with Bootstrap containing the actual elements in a html, then use your library to handle the AJAX implementation. That way I get the ease of your EmbAJAX and the full styling of Bootstrap :-)

The page template idea was to create a way to create a page entirely in one class. Made some progress - 

```
// AJAXWebpage class
// Built on EmbAJAX
template<typename Path, typename ... Elements>
class AJAXWebPage<Path, Elements ...>
{
    public:
        AJAXWebPage(const Path& path, const Elements& ... elements)
            : path(path)
            , elements(elements...)
            , page_elements{const_cast<Elements*>(&elements)...}
            , page(page_elements,"","")
        {}
        
        Path path;
        AJAXWebPage<Elements ... > elements;

        EmbAJAXBase* page_elements[sizeof...(Elements)];
        EmbAJAXPage<sizeof...(Elements)> page;

        template<size_t idx>
        auto get()
        {
            return GetHelper<idx, AJAXWebPage<Path,Elements...>>::get(*this);
        }
};
```

A page can be created like this -

```
EmbAJAXSlider slider("slider", 0, 1000, 500);
const char* radio_opts[] = {"Option1", "Option2", "Option3"};
void buttonPressed(EmbAJAXPushButton*) { return; }

AJAXWebPage
    < const char*
    , EmbAJAXRadioGroup<3>
    , EmbAJAXPushButton
    , EmbAJAXSlider
> page1
    ( "/page1.html"
    , {"radio",radio_opts}
    , {"button", "I can count", buttonPressed}
    , {"slider", 0, 1000, 500}
);
```

Not sure I understand your container - can you give me a quick example of how that would work?
Thanks!

**Comment by tfry-git** [comment]
Hi!

I thought (and think), the biggest problem is: How can you keep a list of your AJAXWebPage-objects, when each of then is technically a distinct class (due to the template parameters)? But perhaps you have some really nifty logic in that ```GetHelper``` to help you around that problem?

Anyway, here's what I had in mind for you. For simplicity I've stripped all template parameters. While I rather like your idea of using variadic args, I've also stripped that out for now, for an easier view on the main idea:

```
class AJAXWebPage {
    public:
        AJAXWebPage(const char* path, size_t num, EmbAJAXBase **elements)
            : path(path)
            , page_elements(num, elements)
            , dummy{&page_elements}
            , page(dummy,"","")
        {}
        
        const char* path;

        EmbAJAXElementList page_elements;
        // I suppose the compiler can be coerced to get along without this dummy, somehow:
        EmbAJAXBase* dummy[1];
        EmbAJAXPage<1> page;

        EmbAJAXBase* get(int index)
        {
            return page_elements.getChild(index);
        }
};
```

The main advantage should be that all your AJAXWebPage instances are the same class, as you have no template parameters (or the remaining template parameters can be the same for all instances). That said, perhaps I still did not quite get your idea. Perhaps a more complete example would help.

**Comment by csgregg** [comment]
Thanks - not quite what I was looking for.  I wanted to still put all the elements into the page class rather that point to them on the outside. 

I tried another approach. I don't mind having different classes per page, as long as I can have a generic request handler, and send the page content from SPIFFS so I can fully style with Bootstrap. Below is where I ended up and it's close enough for that I wanted to do. Thanks for the input and the great library.

BTW - any plans too add the library to Platformio's library manager?

```
struct thiswebpage {

    const char* URL;
    void (*handler)();

    EmbAJAXCheckButton check;
    EmbAJAXMutableSpan check_d;

    EmbAJAXBase* page_elements[2] = {
        &check,
        &check_d
    };

    thiswebpage( const char* pURL, void(*phandler)() ) : 
        check("check", ""),
        check_d("check_d"),
        ajax(page_elements, "")
        {
            URL = pURL;
            handler = phandler;
        };

    EmbAJAXPage<sizeof(page_elements)/sizeof(EmbAJAXBase*)> ajax;

} _thispage("/this.html", []() { 
        _thispage.ajax.handleRequest( []() {
            _thispage.check_d.setValue(_thispage.check.isChecked() ? " Checked" : " Not checked");
        } );
    } );
    

struct thatwebpage {

    const char* URL;
    void (*handler)();

    EmbAJAXCheckButton check;
    EmbAJAXMutableSpan check_d;

    EmbAJAXBase* page_elements[2] = {
        &check, 
        &check_d
    };

    thatwebpage( const char* pURL, void(*phandler)() ) : 
        check("check", ""),
        check_d("check_d"),
        ajax(page_elements, "")
        {
            URL = pURL;
            handler = phandler;
        };

    EmbAJAXPage<sizeof(page_elements)/sizeof(EmbAJAXBase*)> ajax;

} _thatpage("/that.html", []() { 
        _thatpage.ajax.handleRequest( []() {
            _thatpage.check_d.setValue(_thatpage.check.isChecked() ? " Checked" : " Not checked");
        } );
    } );


PageHandler pagehandlers[] = {
    {_thispage.URL, _thispage.handler},
    {_thatpage.URL, _thatpage.handler},
};



void WebsiteManager::handleAJAX( const String path ) {

    for( u_int i = 0; i < sizeof(pagehandlers)/sizeof(PageHandler); i++ )
        if( path == pagehandlers[i].URL ) (pagehandlers[i].handler)();
}


void WebsiteManager::begin() {

     // If the client requests any URI
    _server.onNotFound( 

        [&]() {                             

            // AJAX request
            if( _server.method() == HTTP_POST ) handleAJAX( _server.uri() ); 

            // Page request - send it if it exists otherwise, respond with a 404 (Not Found) error
            else if( !handleSPIFFS( _server.uri() ) ) _server.send( 404, F("text/html"), F("404: Not Found") );
        
        }

    );

    SPIFFS.begin();
    _server.begin(); 

}


String WebsiteManager::getContentType(String filename) { // convert the file extension to the MIME type
    if (filename.endsWith(F(".html.gz"))) return F("text/html");
    else if (filename.endsWith(F(".html"))) return F("text/html");
    else if (filename.endsWith(F(".css.gz"))) return F("text/css");
    else if (filename.endsWith(F(".css"))) return F("text/css");
    else if (filename.endsWith(F(".js.gz"))) return F("application/javascript");
    else if (filename.endsWith(F(".js"))) return F("application/javascript");
    else if (filename.endsWith(F(".ico.gz"))) return F("image/x-icon");
    else if (filename.endsWith(F(".ico"))) return F("image/x-icon");
    else if (filename.endsWith(F(".woff.gz"))) return F("application/font-woff");
    else if (filename.endsWith(F(".woff"))) return F("application/font-woff");
    return F("text/plain");
}


// send the right file to the client (if it exists)
bool WebsiteManager::handleSPIFFS(String shortpath) {

    if (shortpath.endsWith("/")) shortpath += F("index.html");         // If a folder is requested, send the index file

    shortpath.replace(F("fonts/glyphicons-halflings-regular"), F("font"));      // Deal with std bootstrap naming

    String path = shortpath + ".gz";

    String contentType = getContentType(path);              // Get the MIME type
    if( SPIFFS.exists(path) ) {                             // If the file exists then send it
        File file = SPIFFS.open(path, "r");
        _server.streamFile(file, contentType);
        file.close();
        return true;
    }

    return false;                                         // If the file doesn't exist, return false
}
```
